src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/LocalFolderUtility.cs

```diff
@@ -0,0 +1,685 @@
+
+                throw new FatalProtocolException(message, ex);
+            }
+
+            if (!Directory.Exists(rootDirectoryInfo.FullName))
+            {
+                // Directory is missing
+                yield break;
+            }
+
+            var pathResolver = new VersionFolderPathResolver(root);
+            var idRoot = Path.Combine(root, id);
+
+            foreach (var versionDir in GetDirectoriesSafe(idRoot, log))
+            {
+                NuGetVersion version;
+                if (NuGetVersion.TryParse(versionDir.Name, out version))
+                {
+                    var identity = new PackageIdentity(id, version);
+
+                    // Read the package, this may be null if files are missing
+                    var package = GetPackageV3(root, identity, log);
+
+                    if (package != null)
+                    {
+                        yield return package;
+                    }
+                }
+                else
+                {
+                    log.LogWarning(string.Format(CultureInfo.CurrentCulture, Strings.UnableToParseFolderV3Version, versionDir));
+                }
+            }
+
+            yield break;
+        }
+
+        /// <summary>
+        /// Remove duplicate packages which can occur in directories.
+        /// In V2 packages may exist under multiple sub folders. 
+        /// Non-normalized versions also lead to duplicates: ex: 1.0, 1.0.0.0
+        /// </summary>
+        public static IEnumerable<LocalPackageInfo> GetDistinctPackages(IEnumerable<LocalPackageInfo> packages)
+        {
+            if (packages == null)
+            {
+                throw new ArgumentNullException(nameof(packages));
+            }
+
+            var seen = new HashSet<PackageIdentity>();
+
+            foreach (var package in packages)
+            {
+                if (seen.Add(package.Identity))
+                {
+                    yield return package;
+                }
+            }
+
+            yield break;
+        }
+
+        /// <summary>
+        /// Retrieve directories and log exceptions that occur.
+        /// </summary>
+        private static DirectoryInfo[] GetDirectoriesSafe(string root, ILogger log)
+        {
+            try
+            {
+                var rootDir = new DirectoryInfo(root);
+                return rootDir.GetDirectories();
+            }
+            catch (Exception e)
+            {
+                log.LogWarning(e.Message);
+            }
+
+            return new DirectoryInfo[0];
+        }
+
+        /// <summary>
+        /// Retrieve files and log exceptions that occur.
+        /// </summary>
+        private static FileInfo[] GetFilesSafe(string root, string filter, ILogger log)
+        {
+            try
+            {
+                var rootDir = new DirectoryInfo(root);
+                return rootDir.GetFiles(filter);
```

Comment:
Apparently you can't enumerate with try catch, I tried to make it work. These safe methods were taken from nuget.core.dll, it looks like we support scenarios where the user doesn't have access to all directories in a folder.